### ISSUE TITLE
Join `ON` default indentation is behaving differently from 1.4.x version and 2.0.0a5 documentation

### ISSUE THREAD

**Issue by reneoctavio** [sqlfluff/sqlfluff#4429]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

We are testing our codebase with the recent alpha version 2.0.0a5, and the indentation in joins that we currently use in 1.4.5, and that is the default in 2.0.0a5 documentation is not the same. Basically, it is asking for a new line after the `ON` condition.

This is a simple example.
```sql
SELECT
    table1.a,
    table1.b,
    table2.c,
    table3.d
FROM table1
INNER JOIN table2
    ON table1.a = table2.a
        AND table1.b = table2.b
LEFT JOIN table3
    ON table1.a = table3.a

```

The documentation: https://github.com/sqlfluff/sqlfluff/blob/2.0.0a5/docs/source/layout.rst#configuring-indent-locations

### Expected Behaviour

That `sqlfluff lint` should not fail. And this file should not be fixed.

### Observed Behaviour

The test of the lint.
```
== [test.sql] FAIL                                                                                                                                                                    
L:   8 | P:   7 | L003 | Expected line break and indent of 8 spaces before
                       | 'table1'. [layout.indent.b]
All Finished 📜 🎉!
```

The proposed fix:
```sql
SELECT
    table1.a,
    table1.b,
    table2.c,
    table3.d
FROM table1
INNER JOIN table2
    ON
        table1.a = table2.a
        AND table1.b = table2.b
LEFT JOIN table3
    ON table1.a = table3.a

```

### How to reproduce

Basically, run the vanilla `sqlfluff` with the given query.

### Dialect

ansi, databricks

### Version

```
==== sqlfluff ====
sqlfluff:              2.0.0a5 python:                 3.11.1
implementation:        cpython verbosity:                   4
dialect:                  ansi templater:               jinja
rules:                                all
```

### Configuration

```ini
[sqlfluff]
dialect = ansi
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Thanks for the heads up. This is an example of an implicit indent: https://docs.sqlfluff.com/en/latest/layout.html#implicit-indents

They're not enabled by default, but would allow your initial layout. I'll update the documentation accordingly.
